pkg/kubelet/secret/secret_manager.go

```diff
@@ -0,0 +1,272 @@
+	// in other case, this may block whole sync loop in Kubelet.
+	sync.Mutex
+
+	secret         *v1.Secret
+	err            error
+	lastUpdateTime time.Time
+}
+
+// update assumes that the secret and error passed to this function
+// we really want to store (e.g. it is not a transient error).
+// Thus, the only case when we want to ignore it is when the passed
+// secret is older then the one already stored.
+func (s *secretData) update(secret *v1.Secret, err error, timestamp time.Time) {
+	s.Lock()
+	defer s.Unlock()
+	if s.secret != nil && secret != nil && isSecretOlder(secret, s.secret) {
+		return
+	}
+	s.secret = secret
+	s.err = err
+	s.lastUpdateTime = timestamp
+}
+
+// secretStore is a local cache of secrets.
+type secretStore struct {
+	kubeClient clientset.Interface
+
+	lock  sync.Mutex
+	items map[objectKey]*secretStoreItem
+	ttl   time.Duration
+}
+
+func newSecretStore(kubeClient clientset.Interface, ttl time.Duration) *secretStore {
+	return &secretStore{
+		kubeClient: kubeClient,
+		items:      make(map[objectKey]*secretStoreItem),
+		ttl:        ttl,
+	}
+}
+
+func isSecretOlder(newSecret, oldSecret *v1.Secret) bool {
+	if newSecret != nil && oldSecret != nil {
+		newVersion, _ := storageetcd.Versioner.ObjectResourceVersion(newSecret)
+		oldVersion, _ := storageetcd.Versioner.ObjectResourceVersion(oldSecret)
+		if newVersion >= oldVersion {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *secretStore) Add(namespace, name string) {
+	key := objectKey{namespace: namespace, name: name}
+
+	// Add is called from RegisterPod, thus it needs to be efficient.
+	// Thus Add() is only increasing refCount and generation of a given secret.
+	// Then Get() is responsible for fetching if needed.
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	item, exists := s.items[key]
+	if !exists {
+		item = &secretStoreItem{
+			refCount: 0,
+			secret:   &secretData{},
+		}
+		s.items[key] = item
+	}
+
+	item.refCount++
+	// This should trigger fetch on the next Get() operation.
+	item.secret.update(nil, nil, time.Now().Add(-s.ttl))
+}
+
+func (s *secretStore) Delete(namespace, name string) {
+	key := objectKey{namespace: namespace, name: name}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if item, ok := s.items[key]; ok {
+		item.refCount--
+		// TODO: Do we need to trigger fetch on all deletions?
+		if item.refCount == 0 {
+			delete(s.items, key)
+		}
+	}
+}
+
+func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
+	key := objectKey{namespace: namespace, name: name}
+
+	item, shouldFetch := func() (*secretData, bool) {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		if item := s.items[key]; item != nil {
+			data := item.secret
+			data.Lock()
+			defer data.Unlock()
+			return data, data.err != nil || time.Now().After(data.lastUpdateTime.Add(s.ttl))
+		}
+		return nil, false
+	}()
+	if item == nil {
+		return nil, fmt.Errorf("secret %q/%q not registered", namespace, name)
+	}
+
+	if shouldFetch {
```

Comment:
Just realized the problem you mentioned. Yes. I think that what you just suggested sounds ok. Will try to update the PR in few minutes.